File: commonutils/src/main/java/com/jamieadkins/commonutils/ui/BaseRecyclerViewAdapter.java
package com.jamieadkins.commonutils.ui;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.jamieadkins.commonutils.R;

import java.util.ArrayList;
import java.util.List;

/**
 * Holds base logic for recycler view adapters.
 */

public abstract class BaseRecyclerViewAdapter
        extends RecyclerView.Adapter<BaseViewHolder> {
    private List<RecyclerViewItem> mItems;

    public RecyclerViewItem getItemAt(int position) {
        return mItems.get(position);
    }

    public BaseRecyclerViewAdapter() {
        mItems = new ArrayList<>();
    }

    @Override
    public int getItemCount() {
        if (mItems == null) {
            return 0;
        } else {
            return mItems.size();
        }
    }

    @Override
    public int getItemViewType(int position) {
        return mItems.get(position).getItemType();
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case Header.TYPE_HEADER:
                return new HeaderViewHolder(LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.item_header, parent, false));
            case SubHeader.TYPE_SUB_HEADER:
                return new SubHeaderViewHolder(LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.item_subheader, parent, false));
            case GoogleNowSubHeader.TYPE_GOOGLE_NOW_SUB_HEADER:
                return new GoogleNowSubHeaderViewHolder(LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.item_google_subheader, parent, false));
            default:
                throw new RuntimeException("Detail level has not been implemented.");
        }
    }

    @Override
    public void onBindViewHolder(BaseViewHolder holder, int position) {
        holder.bindItem(mItems.get(position));
    }

    public void addItem(RecyclerViewItem item) {
        if (!mItems.contains(item)) {
            mItems.add(item);
            notifyItemInserted(mItems.size() - 1);
        } else {
            updateItem(item);
        }
    }

    public void addItem(int position, RecyclerViewItem item) {
        if (!mItems.contains(item)) {
            mItems.add(position, item);
            notifyItemInserted(position);
        }
    }

    public void replaceItem(int position, RecyclerViewItem newItem) {
        mItems.set(position, newItem);
        notifyItemChanged(position);
    }

    public void updateItem(RecyclerViewItem updatedItem) {
        int index = mItems.indexOf(updatedItem);
        if (index != -1) {
            replaceItem(index, updatedItem);
        } else {
            addItem(updatedItem);
        }
    }

    public void removeItem(RecyclerViewItem itemToRemove) {
        if (mItems.contains(itemToRemove)) {
            int index = mItems.indexOf(itemToRemove);
            removeItem(index);
        }
    }

    public void removeItem(int index) {
        mItems.remove(index);
        notifyItemRemoved(index);
    }

    public boolean isAnItemAt(int index) {
        return mItems.size() > index;
    }

    public void clear() {
        mItems.clear();
        notifyDataSetChanged();
    }

    public List<RecyclerViewItem> getItems() {
        return mItems;
    }
}

File: commonutils/src/main/java/com/jamieadkins/commonutils/ui/BaseRecyclerViewAdapter.kt
package com.jamieadkins.commonutils.ui

import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

import com.jamieadkins.commonutils.R
import com.jamieadkins.commonutils.mvp2.applySchedulers
import io.reactivex.Completable
import io.reactivex.Single

import java.util.ArrayList

/**
 * Holds base logic for recycler view adapters.
 */

abstract class BaseRecyclerViewAdapter : RecyclerView.Adapter<BaseViewHolder>() {
    var items: List<RecyclerViewItem> = mutableListOf()
        set(value) {
            calculateDiffAsync(DiffUtilCallback(items, value))
                    .applySchedulers()
                    .subscribe { diffResult: DiffUtil.DiffResult ->
                        field = value
                        diffResult.dispatchUpdatesTo(this)
                    }
        }

    override fun getItemCount(): Int {
        return items.size
    }

    fun calculateDiffAsync(callback: DiffUtil.Callback) : Single<DiffUtil.DiffResult> {
        return Single.just(DiffUtil.calculateDiff(callback))
    }

    fun addItem(item: RecyclerViewItem) {
        addItem(-1, item)
    }

    fun addItem(position: Int, item: RecyclerViewItem) {
        if (items !is MutableList<RecyclerViewItem>) {
            return
        }

        if (item in items) {
            val index = items.indexOf(item)
            if (!item.areContentsTheSame(items[index])) {
                (items as MutableList<RecyclerViewItem>)[index] = item
                notifyItemChanged(index)
            }
        } else {
            if (position != -1) {
                (items as MutableList<RecyclerViewItem>).add(position, item)
                notifyItemInserted(position)
            } else {
                (items as MutableList<RecyclerViewItem>).add(item)
                notifyItemInserted(itemCount - 1)
            }
        }
    }

    fun removeItem(item: RecyclerViewItem) {
        if (items !is MutableList<RecyclerViewItem>) {
            return
        }

        if (item in items) {
            val index = items.indexOf(item)
            (items as MutableList<RecyclerViewItem>).removeAt(index)
            notifyItemRemoved(index)
        }
    }

    fun removeItemAt(position: Int) {
        if (items !is MutableList<RecyclerViewItem>) {
            return
        }

        (items as MutableList<RecyclerViewItem>).removeAt(position)
        notifyItemRemoved(position)
    }

    override fun getItemViewType(position: Int): Int {
        return items[position].itemType
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        when (viewType) {
            Header.TYPE_HEADER -> return HeaderViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_header, parent, false))
            SubHeader.TYPE_SUB_HEADER -> return SubHeaderViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_subheader, parent, false))
            GoogleNowSubHeader.TYPE_GOOGLE_NOW_SUB_HEADER -> return GoogleNowSubHeaderViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_google_subheader, parent, false))
            else -> throw RuntimeException("Detail level has not been implemented.")
        }
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        holder.bindItem(items[position])
    }
}
